src/parsers/xml-processor/utf8.php

```diff
@@ -0,0 +1,363 @@
+	if ( $byte_offset < 0 ) {
+		return null;
+	}
+
+	$position_in_input = $byte_offset;
+	$code_point_at     = $byte_offset;
+	$end_byte          = strlen( $text );
+	$codepoint         = null;
+	$decoder_state     = UTF8_DECODER_ACCEPT;
+
+	// Get to the start of the string.
+	while ( $position_in_input < $end_byte ) {
+		$decoder_state = utf8_decoder_apply_byte( $text[ $position_in_input ], $decoder_state );
+
+		if ( UTF8_DECODER_ACCEPT === $decoder_state ) {
+			++$position_in_input;
+			$codepoint = utf8_ord( substr( $text, $code_point_at, $position_in_input - $code_point_at ) );
+			break;
+		} elseif ( UTF8_DECODER_REJECT === $decoder_state ) {
+			// "\u{FFFD}" is not supported in PHP 5.6.
+			$codepoint = utf8_ord( "\xEF\xBF\xBD" );
+			break;
+		} else {
+			++$position_in_input;
+		}
+	}
+
+	$matched_bytes = $position_in_input - $byte_offset;
+
+	return $codepoint;
+}
+
+/**
+ * Convert a UTF-8 byte sequence to its Unicode codepoint.
+ *
+ * @param  string  $character  UTF-8 encoded byte sequence representing a single Unicode character.
+ *
+ * @return int Unicode codepoint.
+ */
+function utf8_ord( $character ) {
+	// Convert the byte sequence to its binary representation
+	$bytes = unpack( 'C*', $character );
+
+	// Initialize the codepoint
+	$codepoint = 0;
+
+	// Calculate the codepoint based on the number of bytes
+	if ( count( $bytes ) === 1 ) {
+		$codepoint = $bytes[1];
+	} elseif ( count( $bytes ) === 2 ) {
+		$codepoint = ( ( $bytes[1] & 0x1F ) << 6 ) | ( $bytes[2] & 0x3F );
+	} elseif ( count( $bytes ) === 3 ) {
+		$codepoint = ( ( $bytes[1] & 0x0F ) << 12 ) | ( ( $bytes[2] & 0x3F ) << 6 ) | ( $bytes[3] & 0x3F );
+	} elseif ( count( $bytes ) === 4 ) {
+		$codepoint = ( ( $bytes[1] & 0x07 ) << 18 ) | ( ( $bytes[2] & 0x3F ) << 12 ) | ( ( $bytes[3] & 0x3F ) << 6 ) | ( $bytes[4] & 0x3F );
+	}
+
+	return $codepoint;
+}
+
+/**
+ * Encode a code point number into the UTF-8 encoding.
+ *
+ * This encoder implements the UTF-8 encoding algorithm for converting
+ * a code point into a byte sequence. If it receives an invalid code
+ * point it will return the Unicode Replacement Character U+FFFD `�`.
+ *
+ * Example:
+ *
+ *     '🅰' === WP_HTML_Decoder::code_point_to_utf8_bytes( 0x1f170 );
+ *
+ *     // Half of a surrogate pair is an invalid code point.
+ *     '�' === WP_HTML_Decoder::code_point_to_utf8_bytes( 0xd83c );
+ *
+ * @since 6.6.0
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc3629 For the UTF-8 standard.
+ *
+ * @param int $code_point Which code point to convert.
+ * @return string Converted code point, or `�` if invalid.
+ */
+function code_point_to_utf8_bytes( $code_point ) {
+	// Pre-check to ensure a valid code point.
+	if (
+		$code_point <= 0 ||
+		( $code_point >= 0xD800 && $code_point <= 0xDFFF ) ||
+		$code_point > 0x10FFFF
+	) {
+		return '�';
+	}
+
+	if ( $code_point <= 0x7F ) {
+		return chr( $code_point );
+	}
+
+	if ( $code_point <= 0x7FF ) {
+		$byte1 = chr( ( $code_point >> 6 ) | 0xC0 );
```

Comment:
This is similar to [what I do](https://github.com/zaerl/mojibake/blob/main/src/encoding.c#L196), and others as well. What we have here is working because we are checking U+0080–U+07FF.

The mask I use is still considered a good practice: it documents that you want to “keep 5 bits” and not the rest of the integer. Other checks that follow are the same.